### PR TITLE
New algorithm for autobinning

### PR DESCRIPTION
**Pull request by gganis**
This implements a proof of concept of a new histogram auto binning mode designed to produced mergeable histograms when run in multi-threaded / multi-process mode.

The algorithm chooses powers of 2 to chose the boundaries, ranges and number of bins. The idea is described in the following document:  https://docs.google.com/document/d/1TEntRbVnRzZRLio8JhND51FTNqqTAvu0ngnhLGneahg/edit?usp=sharing

A method to adjust the final looking of the histogram is also included in the patch.

Two multicore tutorials illustrate the usage: mt303_fillHistosAutobin.C and mt301_fillHistos.C .

### CODE REVIEW

File: hist/hist/inc/TH1.h [hunk 1]
```diff
@@ -145,14 +152,19 @@ class TH1 : public TNamed, public TAttLine, public TAttFill, public TAttMarker {
 public:
    // TH1 status bits
    enum EStatusBits {
-      kNoStats     = BIT(9),  ///< don't draw stats box
+      kNoStats = BIT(9),      ///< don't draw stats box
```
Comment (pcanal) [inline hist/hist/inc/TH1.h]: Please don't listen to clang-format for this enum :)

File: hist/hist/src/TH1.cxx [hunk 2]
```diff
@@ -566,6 +567,9 @@ TH1::TH1(): TNamed(), TAttLine(), TAttFill(), TAttMarker()
    fXaxis.SetParent(this);
    fYaxis.SetParent(this);
    fZaxis.SetParent(this);
+   ResetBit(TH1::kAutoBinPTwo);
```
Comment (pcanal) [inline hist/hist/src/TH1.cxx]: What is the purpose of resetting them here?

Related question.  Are those bits intended to be persistent or transient?

File: math/mathcore/inc/TStatistic.h [hunk 3]
```diff
@@ -55,8 +52,10 @@ class TStatistic : public TObject {
    inline       Long64_t GetN() const { return fN; }
    inline       Long64_t GetNeff() const { return fW*fW/fW2; }
    inline       Double_t GetM2() const { return fM2; }
+   inline Double_t GetMax() const { return fMax; }
```
Comment (pcanal) [inline math/mathcore/inc/TStatistic.h]: Please align with the function names :)

File: tutorials/multicore/mt303_fillHistosAutobin.C [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+#include "TH1D.h"
+#include "TMath.h"
+#include "TF1.h"
+#include "TCanvas.h"
+#include "TRandom3.h"
+#include "TStatistic.h"
+#include "TFile.h"
+#include "TStyle.h"
+
+TF1 *gam = new TF1("gam", "1/(1+0.1*x*0.1*x)", -100., 100.);
```
Comment (pcanal) [inline tutorials/multicore/mt303_fillHistosAutobin.C]: Do those function (pointer) really need to be global?

File: math/mathcore/inc/TStatistic.h (hunk 3 above)
Comment (gganis) [reply math/mathcore/inc/TStatistic.h]: This is clang-format. Ok, I will ignore these.

File: hist/hist/inc/TH1.h [hunk 5]
```diff
@@ -96,6 +98,7 @@ class TH1 : public TNamed, public TAttLine, public TAttFill, public TAttMarker {
     TList        *fFunctions;       ///<->Pointer to list of functions (fits and user)
     Int_t         fBufferSize;      ///< fBuffer size
     Double_t     *fBuffer;          ///<[fBufferSize] entry buffer
+    TStatistic fXstat;              ///< Statistic about X in fBuffer (max, min, mean, rms, ...)
```
Comment (pcanal) [inline hist/hist/inc/TH1.h]: This increases the size of all histogram by 5 doubles.  Isn't any of the field redudant (i.e. is there other field than can be dropped)?

File: tutorials/multicore/mt303_fillHistosAutobin.C (hunk 4 above)
Comment (gganis) [reply tutorials/multicore/mt303_fillHistosAutobin.C]: No. This was a working macro, not really intended to make it to the end ... at least without proper revision.

File: hist/hist/src/TH1.cxx [hunk 6]
```diff
@@ -1221,6 +1229,222 @@ void TH1::AddDirectory(Bool_t add)
...
 ////////////////////////////////////////////////////////////////////////////////
+/// Auxilliary function to get the next power of 2 value larger then x
+///
+/// Used by the autobin power of 2 algorithm
+
+Double_t TH1::AutoP2GetMax(Double_t x)
+{
+   Double_t neg = (x < 0) ? -1. : 1.;
+   Double_t y = neg * x;
+   Double_t xn = TMath::Log2(y);
```
Comment (pcanal) [inline hist/hist/src/TH1.cxx]: Please consider use std::frexp (very fast) if possible:
```
   double fraction = std::frexp(arg, &exp);
   // frexp doc is:
   //    if no errors occur,
   //    returns the value x in the range (-1;-0.5], [0.5; 1)
   //    and stores an integer value in *exp such that x×2^(*exp) = arg
```

File: hist/hist/inc/TH1.h (hunk 5 above)
Comment (gganis) [reply hist/hist/inc/TH1.h]: Yes, this can be done differently, with at most two additonial doubles.

File: hist/hist/src/TH1.cxx (hunk 6 above)
Comment (gganis) [reply hist/hist/src/TH1.cxx]: ok, I'll do

**Comment by gganis** [comment]
Revised and simplified in #1305 .

**Pull request closed without merging**